pkg/kubelet/deviceplugin/manager.go

```diff
@@ -1,236 +0,0 @@
-// f is the callback that is called when a device becomes unhealthy.
-// socketPath is present for testing purposes in production this is pluginapi.KubeletSocket
-func NewManagerImpl(socketPath string, f MonitorCallback) (*ManagerImpl, error) {
-	glog.V(2).Infof("Creating Device Plugin manager at %s", socketPath)
-
-	if socketPath == "" || !filepath.IsAbs(socketPath) {
-		return nil, fmt.Errorf(errBadSocket+" %v", socketPath)
-	}
-
-	dir, file := filepath.Split(socketPath)
-	return &ManagerImpl{
-		Endpoints: make(map[string]*endpoint),
-
-		socketname: file,
-		socketdir:  dir,
-		callback:   f,
-	}, nil
-}
-
-func (m *ManagerImpl) removeContents(dir string) error {
-	d, err := os.Open(dir)
-	if err != nil {
-		return err
-	}
-	defer d.Close()
-	names, err := d.Readdirnames(-1)
-	if err != nil {
-		return err
-	}
-	for _, name := range names {
-		filePath := filepath.Join(dir, name)
-		if filePath == m.CheckpointFile() {
-			continue
-		}
-		stat, err := os.Stat(filePath)
-		if err != nil {
-			glog.Errorf("Failed to stat file %v: %v", filePath, err)
-			continue
-		}
-		if stat.IsDir() {
-			continue
-		}
-		err = os.RemoveAll(filePath)
-		if err != nil {
-			return err
-		}
-	}
-	return nil
-}
-
-// CheckpointFile returns device plugin checkpoint file path.
-func (m *ManagerImpl) CheckpointFile() string {
-	return filepath.Join(m.socketdir, "kubelet_internal_checkpoint")
-}
-
-// Start starts the Device Plugin Manager
-func (m *ManagerImpl) Start() error {
-	glog.V(2).Infof("Starting Device Plugin manager")
-
-	socketPath := filepath.Join(m.socketdir, m.socketname)
-	os.MkdirAll(m.socketdir, 0755)
-
-	// Removes all stale sockets in m.socketdir. Device plugins can monitor
-	// this and use it as a signal to re-register with the new Kubelet.
-	if err := m.removeContents(m.socketdir); err != nil {
-		glog.Errorf("Fail to clean up stale contents under %s: %+v", m.socketdir, err)
-	}
-
-	if err := os.Remove(socketPath); err != nil && !os.IsNotExist(err) {
-		glog.Errorf(errRemoveSocket+" %+v", err)
-		return err
-	}
-
-	s, err := net.Listen("unix", socketPath)
-	if err != nil {
-		glog.Errorf(errListenSocket+" %+v", err)
-		return err
-	}
-
-	m.server = grpc.NewServer([]grpc.ServerOption{}...)
-
-	pluginapi.RegisterRegistrationServer(m.server, m)
-	go m.server.Serve(s)
-
-	return nil
-}
-
-// Devices is the map of devices that are known by the Device
-// Plugin manager with the kind of the devices as key
-func (m *ManagerImpl) Devices() map[string][]*pluginapi.Device {
-	m.mutex.Lock()
-	defer m.mutex.Unlock()
-
-	devs := make(map[string][]*pluginapi.Device)
-	for k, e := range m.Endpoints {
-		glog.V(3).Infof("Endpoint: %+v: %+v", k, e)
-		devs[k] = e.getDevices()
-	}
-
-	return devs
-}
-
-// Allocate is the call that you can use to allocate a set of devices
-// from the registered device plugins.
-func (m *ManagerImpl) Allocate(resourceName string, devs []string) (*pluginapi.AllocateResponse, error) {
-
-	if len(devs) == 0 {
```

Comment:
Looks like this check doesnt show up in DevicePluginHandlerImpl.Allocate.  Is it safe to drop this check?

Reply:
Should be, the allocation action of specific device plugins is supposed to be actions in a range of the devs, the return directly if devs empty.
Anyway, I would prefer add the check back for safety :)